### DISCUSSION TITLE
Got error when posting `admin get engine` REST with `sharelevel=GROUP`

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#4684] [Q&A]
Hi Kyuubi community,

I have created 2 Spark-cluster with same user `remziy`. The former is created in `USER` share level and the latter is created in `GROUP` share level (primary_group = "group_remziy"). I can get the info of the former using the `admin get engine` REST. On the other hand, I got the error `KeeperErrorCode = NoNode for /kyuubi_1.7.0_GROUP_SPARK_SQL/remziy` when getting access to the latter. Any idea of how to fix it?

**Comment by HaoYang670** [comment]
@zwangsheng Could you please take a look if you have time? Thanks a lot.

**Reply by zwangsheng** [reply]
Would you like show some params, such like Kyuubi Server Configurations about `ShareLevel` or `kyuubi.engine.share.level.subdomain`?

BTW, you should parse rest url with request params for giving more info.

Also, if `GROUP level` spark engine had been created, you can find engine registed path from spark driver log.

**Reply by HaoYang670** [reply]
> Would you like show some params?

Sure. We set the `sharelevel` in the session properties of JDBC URL and leave the `subdomain` as `default`. 

> BTW, you should parse rest url with request params for giving more info.

The rest URL we send is `http://{{KYUUBI_HOST}}:{{REST_PORT}}/api/v1/admin/engine?sharelevel=GROUP`. And we have updated the authentication header to match the username of the kyuubi engine.

**Reply by HaoYang670** [reply]
BTW, I can get the handlers of kyuubi sessions of the cluster, and get the info of the kyuubi session. 
```json
{
    "sessionId": "*",
    "clientVersion": 9,
    "sessionType": "INTERACTIVE",
    "sessionName": "",
    "user": "remziy",
    "clientIP": "10.244.0.1",
    "serverIP": "0.0.0.0:10009",
    "conf": {
        "kyuubi.session.real.user": "remziy",
        "set:hiveconf:spark.executor.cores": "1",
        "set:hiveconf:spark.hadoop.fs.s3a.secret.key": "*",
        "set:hiveconf:spark.executor.memory": "512m",
        "kyuubi.client.version": "1.7.0",
        "use:database": "default",
        "set:hiveconf:spark.kubernetes.container.image": "*",
        "set:hiveconf:kyuubi.engine.share.level": "GROUP",
        "set:hiveconf:spark.executor.instances": "1",
        "kyuubi.client.ipAddress": "*",
        "set:hiveconf:hive.server2.thrift.resultset.default.fetch.size": "1000",
        "set:hiveconf:spark.hadoop.fs.s3a.access.key": "*",
        "kyuubi.session.connection.url": "0.0.0.0:10009",
        "kyuubi.server.ipAddress": "0.0.0.0",
        "set:hiveconf:spark.kubernetes.namespace": "default"
    },
    "eventTime": 1681202618402,
    "startTime": 1681202618401,
    "remoteSessionId": "*",
    "engineId": "*",
    "openedTime": 1681202618415,
    "endTime": -1,
    "totalOperations": 0,
    "exception": null,
    "eventType": "kyuubi_session"
}

**Comment by zwangsheng** [comment]
Can you provide spark driver log about `Registered EngineServiceDiscovery in namespace ${_namespace}`, which will help find where sparkSqlEngine registered.

**Reply by HaoYang670** [reply]
Sorry for the late response.

> Can you provide spark driver log about Registered EngineServiceDiscovery in namespace ${_namespace}, 

Sure. This is what I got in my Spark driver log:
```bash
$ kubectl logs kyuubi-group-spark-sql-... | grep EngineService
23/04/17 01:46:50 INFO EngineServiceDiscovery: Service[EngineServiceDiscovery] is initialized.
23/04/17 01:46:50 INFO EngineServiceDiscovery: Service[EngineServiceDiscovery] is started.
```

**Reply by HaoYang670** [reply]
BTW. Are there other APIs that can get the engine information?

**Reply by zwangsheng** [reply]
Currently, kyuubi only provide `/api/admin/engines` for get engine info by restful api.

You can find all rest api in https://kyuubi.readthedocs.io/en/v1.7.0/client/rest/rest_api.html#admin-resource

Also cc @turboFei @lightning-L

**Reply by HaoYang670** [reply]
Oh, I find the engine is in the namespace `/kyuubi_1.7.0_GROUP_SPARK_SQL/group_remziy`, where `group_remziy` is the primary group of `remziy`. But it should be the user name isn't it?

**Reply by zwangsheng** [reply]
Find some logic in branch 1.7:
```
* The EngineSpace used to expose itself to the KyuubiServers in `serverSpace`
   *
   * For `CONNECTION` share level:
   *   /`serverSpace_version_CONNECTION_engineType`/`user`/`engineRefId`
   * For `USER` share level:
   *   /`serverSpace_version_USER_engineType`/`user`[/`subdomain`]
   * For `GROUP` share level:
   *   /`serverSpace_version_GROUP_engineType`/`primary group name`[/`subdomain`]
   * For `SERVER` share level:
   *   /`serverSpace_version_SERVER_engineType`/`kyuubi server user`[/`subdomain`]
 ```

Seems. this is a bug, `AdminResource# list enigne`'s logic is inconsistent with `EngineRef`'s.

**Reply by HaoYang670** [reply]
I think for GROUP level, we should use
```
/`serverSpace_version_GROUP_engineType`/`user`[/`subdomain`]
```

right?

**Reply by zwangsheng** [reply]
Thanks for the reminder, if you are happy to open an issue, I will fix this (of course you are welcome to contribute if you want).

**Reply by pan3793** [reply]
It's by design to use `primary group name`
```
/`serverSpace_version_GROUP_engineType`/`primary group name`[/`subdomain`]
```

**Reply by HaoYang670** [reply]
> It's by design to use primary group name

Then, how could I use the admin RESTful API to get info of an engine in GROUP share level?

**Reply by zwangsheng** [reply]
>I think for GROUP level, we should use
/`serverSpace_version_GROUP_engineType`/`user`[/`subdomain`]
right?

Yes

In AdminResource, will find GROUP LEVEL engine in `/serverSpace_version_GROUP_engineType/user`, but engine register self in `/serverSpace_version_GROUP_engineType/prime_user`

We should keep `AdminResource#getEngineSpace` logic same with `EngineRef#engineSpace`

**Reply by HaoYang670** [reply]
The admin RESTful API doesn't convert my username to the primary group name, which is why I always got `KeeperNoNodeError`

**Reply by HaoYang670** [reply]
> In AdminResource, will find GROUP LEVEL engine in /serverSpace_version_GROUP_engineType/user, but engine register self in /serverSpace_version_GROUP_engineType/prime_user

I think this is a bug, right?

**Reply by zwangsheng** [reply]
> > In AdminResource, will find GROUP LEVEL engine in /serverSpace_version_GROUP_engineType/user, but engine register self in /serverSpace_version_GROUP_engineType/prime_user
> 
> I think this is a bug, right?

Yes, but we should keep as `/serverSpace_version_GROUP_engineType/prime_user`, so we should modify `AdminResource#getEngineSpace`

**Reply by HaoYang670** [reply]
File an issue to track it: https://github.com/apache/kyuubi/issues/4738
We can move the discussion there
